File: scala/src/main/scala/com/shekhargulati/ninetynine_problems/_01_lists/P15.scala
package com.shekhargulati.ninetynine_problems._01_lists

/**
  * <b>(**) Duplicate the elements of a list a given number of times.</b>
  *
  * {{{
  *   duplicateTimes(List("a", "b", "c"), 3)
  * }}}
  */
object P15 {

  /*
  Using flatMap and List.fill
   */
  def duplicateTimes[T](list: List[T], times: Int): List[T] = list.flatMap(x => List.fill(times)(x))


  /*
  Using recursion
   */
  def duplicateTimes_recursion[T](list: List[T], times: Int): List[T] = list match {
    case x :: xs => List.fill(times)(x) ++ duplicateTimes_recursion(xs, times)
    case Nil => Nil
  }

}

File: java8/src/main/java/com/shekhargulati/ninetynine_problems/_01_lists/P15.java
package com.shekhargulati.ninetynine_problems._01_lists;


import java.util.Collections;
import java.util.List;

import static java.util.stream.Collectors.toList;

/**
 * (**) Duplicate the elements of a list a given number of times.
 * <pre>
 *          duplicate(Arrays.asList("a", "b", "c", "d"))
 * </pre>
 */
public class P15 {

    public static <T> List<T> duplicate(final List<T> list, final int times) {
        return list.stream().flatMap(e -> Collections.nCopies(times, e).stream()).collect(toList());
    }
}
